File: backend_fastapi/app/core/receipts.py
import hashlib
import hmac
import json
from datetime import datetime, timezone
from typing import Any

from app.core.policy import make_id, now_iso


def _stable(value: Any) -> str:
    if isinstance(value, list):
        return "[" + ",".join(_stable(v) for v in value) + "]"
    if isinstance(value, dict):
        items = sorted(value.items())
        return "{" + ",".join(f"{json.dumps(k)}:{_stable(v)}" for k, v in items) + "}"
    return json.dumps(value, sort_keys=True)


def sha256(value: Any) -> str:
    if not isinstance(value, str):
        value = _stable(value)
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def sign(value: Any, secret: str) -> str:
    return hmac.new(secret.encode("utf-8"), _stable(value).encode("utf-8"), hashlib.sha256).hexdigest()


def verify(value: Any, signature: str, secret: str) -> bool:
    expected = sign(value, secret)
    return len(signature) == len(expected) and hmac.compare_digest(expected, signature)


def create_receipt(
    *,
    request: dict[str, Any],
    permit: dict[str, Any],
    evaluation: dict[str, Any],
    execution: dict[str, Any] | None = None,
    previous_hash: str = "GENESIS",
    secret: str,
) -> dict[str, Any]:
    body = {
        "id": make_id("rcpt"),
        "workspaceId": permit.get("workspaceId") or permit.get("workspace_id"),
        "permitId": permit.get("id"),
        "agentId": request.get("agentId") or request.get("agent_id"),
        "scope": request.get("scope", ""),
        "target": request.get("target"),
        "amount": float(request.get("amount", 0)),
        "result": evaluation.get("decision", "unknown"),
        "code": evaluation.get("code", ""),
        "reason": evaluation.get("reason", ""),
        "execution": {"status": execution.get("status"), "digest": sha256(execution.get("body", "") or "")} if execution else None,
        "previousHash": previous_hash,
        "createdAt": now_iso(),
    }
    h = sha256(body)
    body["hash"] = h
    body["signature"] = sign({"hash": h, "workspaceId": body["workspaceId"]}, secret)
    return body


def verify_receipt(receipt: dict[str, Any], secret: str) -> bool:
    r = dict(receipt)
    h = r.pop("hash", "")
    sig = r.pop("signature", "")
    body_hash = sha256(r)
    return body_hash == h and verify({"hash": body_hash, "workspaceId": r.get("workspaceId", "")}, sig, secret)

File: packages/core/receipts.mjs
import crypto from 'node:crypto';
import { id, nowIso } from './policy.mjs';

function stable(value) {
  if (Array.isArray(value)) return `[${value.map(stable).join(',')}]`;
  if (value && typeof value === 'object') return `{${Object.keys(value).sort().map(k => `${JSON.stringify(k)}:${stable(value[k])}`).join(',')}}`;
  return JSON.stringify(value);
}

export function sha256(value) { return crypto.createHash('sha256').update(typeof value === 'string' ? value : stable(value)).digest('hex'); }
export function sign(value, secret) { return crypto.createHmac('sha256', secret).update(stable(value)).digest('hex'); }
export function verify(value, signature, secret) {
  const expected = sign(value, secret);
  return signature?.length === expected.length && crypto.timingSafeEqual(Buffer.from(signature), Buffer.from(expected));
}

export function createReceipt({ request, permit, evaluation, execution, previousHash = 'GENESIS', secret, clock = Date }) {
  const body = {
    id: id('rcpt'),
    workspaceId: permit.workspaceId,
    permitId: permit.id,
    agentId: request.agentId,
    scope: request.scope,
    target: request.target || null,
    amount: Number(request.amount || 0),
    result: evaluation.decision,
    code: evaluation.code,
    reason: evaluation.reason,
    execution: execution ? { status: execution.status, digest: sha256(execution.body || '') } : null,
    previousHash,
    createdAt: nowIso(clock)
  };
  const hash = sha256(body);
  return { ...body, hash, signature: sign({ hash, workspaceId: body.workspaceId }, secret) };
}

export function verifyReceipt(receipt, secret) {
  const { hash, signature, ...body } = receipt;
  return hash === sha256(body) && verify({ hash, workspaceId: body.workspaceId }, signature, secret);
}
